### PR TITLE
Viber public handlers

### PR DESCRIPTION
**Pull request by norkans7**

### CODE REVIEW

File: handlers/viber/viber.go [hunk 1]
```diff
@@ -10,3 +27,280 @@ POST /handlers/viber_public/uuid?sig=sig
...
+	confAuth := msg.Channel().ConfigForKey(courier.ConfigAuthToken, "")
+	authToken, isStr := confAuth.(string)
+	if !isStr || authToken == "" {
+		return nil, fmt.Errorf("invalid auth token config")
+	}
+
+	viberMsg := viberOutgoingMessage{
+		AuthToken:    authToken,
+		Receiver:     msg.URN().Path(),
+		Text:         courier.GetTextAndAttachments(msg),
```
Comment (nicpottier) [inline handlers/viber/viber.go]: Whats the max message length? Does viber not support attachments?

Comment (norkans7) [reply handlers/viber/viber.go]: I added the max length, for attachement we can easily support picture as attachment but for video we need to specify the size of the attachment https://developers.viber.com/docs/api/rest-bot-api/#video-message 
I think we need first a way to get that do we fetch first the mediaURL to get the size?

File: handlers/viber/viber.go [hunk 2]
```diff
@@ -10,3 +31,386 @@ POST /handlers/viber_public/uuid?sig=sig
...
+}
+
+// see https://developers.viber.com/docs/api/rest-bot-api/#callbacks
+func (h *handler) validateSignature(channel courier.Channel, r *http.Request) error {
+	actual := r.Header.Get(viberSignatureHeader)
+	if actual == "" {
+		return fmt.Errorf("missing request signature")
+	}
+
+	confAuth := channel.ConfigForKey(courier.ConfigAuthToken, "")
```
Comment (nicpottier) [inline handlers/viber/viber.go]: `channel.StringConfigForKey` will save you a step.

File: handlers/viber/viber.go [hunk 3]
```diff
@@ -10,3 +31,386 @@ POST /handlers/viber_public/uuid?sig=sig
...
+	authToken, isStr := confAuth.(string)
+	if !isStr || authToken == "" {
+		return fmt.Errorf("invalid or missing auth token in config")
+	}
+
+	// read our body
+	body, err := ioutil.ReadAll(r.Body)
+	r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	expected, err := viberCalculateSignature(authToken, body)
```
Comment (nicpottier) [inline handlers/viber/viber.go]: no need for the `viber` prefix, it is a private method

File: handlers/viber/viber.go [hunk 4]
```diff
@@ -10,3 +31,386 @@ POST /handlers/viber_public/uuid?sig=sig
...
+		return fmt.Errorf("invalid request signature")
+	}
+
+	return nil
+}
+
+func viberCalculateSignature(authToken string, contents []byte) ([]byte, error) {
+
+	var buffer bytes.Buffer
+	buffer.Write(contents)
```
Comment (nicpottier) [inline handlers/viber/viber.go]: what purpose does buffer serve? can't we just write the contents directly below?

File: handlers/viber/viber.go [hunk 5]
```diff
@@ -10,3 +31,386 @@ POST /handlers/viber_public/uuid?sig=sig
...
+	AuthToken    string            `json:"auth_token"`
+	Receiver     string            `json:"receiver"`
+	Text         string            `json:"text"`
+	Type         string            `json:"type"`
+	TrackingData string            `json:"tracking_data"`
+	Sender       map[string]string `json:"sender,omitempty"`
+	Media        string            `json:"media,omitempty"`
+	Size         int               `json:"size,omitempty"`
+	FileName     string            `json:"file_name,omitempty"`
+	Keyboard     string            `json:"keyboard,omitempty"`
```
Comment (nicpottier) [inline handlers/viber/viber.go]: This should be a pointer to `viberKeyboard`, not a string. Right now it is going to turn into a double encoded JSON blob

File: utils/http.go [hunk 6]
```diff
@@ -16,14 +16,15 @@ type RequestResponseStatus string
...
-	Body       []byte
-	Elapsed    time.Duration
+	Method        string
+	URL           string
+	Status        RequestResponseStatus
+	StatusCode    int
+	Request       string
+	Response      string
+	Body          []byte
+	ContentLength string
```
Comment (nicpottier) [inline utils/http.go]: Feels like this should be an int (perhaps with a value of -1 if unset)

File: handlers/viber/viber.go (hunk 5 above)
Comment (norkans7) [reply handlers/viber/viber.go]: that way the omitempty is not effective and there is keyboard: {} in the JSON

File: channel_event.go [hunk 7]
```diff
@@ -13,6 +13,7 @@ type ChannelEventType string
 const (
 	NewConversation ChannelEventType = "new_conversation"
 	Referral        ChannelEventType = "referral"
+	StopContact     ChannelEventType = "stop_contact"
```
[the remark this answers is not in the document]
Comment (norkans7) [reply channel_event.go]: added in https://github.com/nyaruka/rapidpro/pull/1688

File: utils/http.go (hunk 6 above)
Comment (norkans7) [reply utils/http.go]: not sure how to set -1 in that case.. should I change RequestResponse{} to always be RequestResponse{ContentLength: -1}?

File: handlers/viber/viber.go (hunk 5 above)
Comment (nicpottier) [reply handlers/viber/viber.go]: You make it a pointer to a `viberKeyboard`, omit empty will leave it out then.

IE: 

`Keyboard *viberKeyboard`

File: utils/http.go (hunk 6 above)
Comment (nicpottier) [reply utils/http.go]: Yes, and should stay -1 if there is no ContentLength header

File: handlers/viber/viber_test.go [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+			}
+		}
+		testCases[c].RequestBody = strings.Replace(testCases[c].RequestBody, "{{ SERVER_URL }}", server.URL, -1)
+	}
+
+	return server
+}
+
+var defaultSendTestCases = []ChannelSendTestCase{
+	{Label: "Plain Send",
```
Comment (nicpottier) [inline handlers/viber/viber_test.go]: Should add a QuickReplies case.

File: handlers/viber/viber.go (hunk 5 above)
Comment (norkans7) [reply handlers/viber/viber.go]: I have tried this and I keep getting {} in the JSON

Comment (norkans7) [reply handlers/viber/viber.go]: never mind, this is now updated

**Pull request merged**